0000-error-values-proposal.rst

```diff
@@ -0,0 +1,282 @@
+  language extensions, importing a module, fixing a typo in the spelling of
+  a name).
+
+* Display squiggles of different colors depending on the nature of the error
+  (e.g "not in scope" errors in some color, "couldn't match" errors in another,
+  and so on, or perhaps discriminating in larger groups). This could help
+  Haskell developers focus on one of those groups (e.g typos in names,
+  missing imports) before proceeding with the more subtle type errors.
+
+* An IDE tool might ask GHC to defer expensive analyses typically done
+  during error message construction (e.g. `computing valid hole fits
+  <https://gitlab.haskell.org/ghc/ghc/issues/16875#note_210045>`_) and instead
+  query GHC for the analysis result asynchronously (or even only when
+  requested by the user), shrinking the edit/typechecking iteration time.
+
+Costs and drawbacks
+-------------------
+
+The import cycles induced by this architecture and the ``.hs-boot`` files that
+we will use to work around them are going to add some maintenance overhead
+which we believe is largely compensated by having errors become proper values.
+
+Another drawback is that the wrapping in ``ParserError``, ``GHCError`` and
+friends can be a bit verbose, becoming more verbose still as we introduce
+additional levels to the hierarchy. This can be mitigated by going for a
+rather flat hierarchy like the one presented above, with ``GHCError`` at the
+top, the subsystem-specific error types below and nothing else: the different
+constructors of those subsystem-specific error types would contain all the
+relevant information for a specific error that GHC can emit, the indirection
+would stop there. This option seems to be a good compromise and that is why
+it is the one we used in the previous section.
+
+The major cost of implementing this proposal is the sheer amount of
+refactoring that will be necessary to emit error values and move the
+rendering to much later, essentially delegating this work to each subystem
+and combining everything in the implementation of ``errorMessage``.
+
+Alternatives
+------------
+
+We considered open variants of this design, where we do not build sum types
+all the way up to ``GHCError``, but where ``GHCError`` is an
+existential wrapper around an open union of error types that provide
+suitable instances:::
+
+    class HasErrMsg e where
+      errorMessage :: e -> ErrMsg
+
+    data GHCError where
+      GHCError :: (Typeable e, HasErrMsg e) => e -> GHCError
+
+This would allow us to work around the whole import cycles problem,
+at the price of being a lot more cumbersome to use: error consumers would
+have to use ``Typeable`` to implement specific behaviours for some types of
+errors. This price is likely a higher one to pay in the long run than the
+import cycles that we would work around when implementing the current proposal,
+as the cost will likely be non-trivial when/if we implement the proposal, but
```

Comment:
Again, I think this cost will be low at all points in time (except maybe when we want to package-ify GHC).